gtsfm/new_detector.py

```diff
@@ -0,0 +1,101 @@
+"""Detector for the front end.
+
+Authors: Ayush Baid
+"""
+import abc
+import pickle
+from pathlib import Path
+
+import dask
+from dask.distributed import Client, LocalCluster, performance_report
+import numpy as np
+import torch
+
+import gtsfm.utils.images as image_utils
+from gtsfm.common.image import Image
+from gtsfm.common.keypoints import Keypoints
+from gtsfm.loader.olsson_loader import OlssonLoader
+from gtsfm.ui.gtsfm_process import GTSFMProcess, UiMetadata
+from thirdparty.SuperGluePretrainedNetwork.models.superpoint import SuperPoint
+
+ROOT_PATH = Path(__file__).resolve().parent.parent
+MODEL_WEIGHTS_PATH = (
+    ROOT_PATH / "thirdparty" / "SuperGluePretrainedNetwork" / "models" / "weights" / "superpoint_v1.pth"
+)
+TEST_DATA_PATH = ROOT_PATH / "tests" / "data" / "set1_lund_door"
+
+
+class NewDetector(GTSFMProcess):
+    @staticmethod
+    def get_ui_metadata() -> UiMetadata:
+        """Returns data needed to display node and edge info for this process in the process graph."""
+
+        return UiMetadata(
+            display_name="NewDetector",
+            input_products=("Images",),
+            output_products=("Keypoints",),
+            parent_plate="DetDescCorrespondenceGenerator",
+        )
+
+    @abc.abstractmethod
+    def apply(self, image: Image) -> Keypoints:
+        """Apply the detector on the image to produce keypoints.
+
+        Args:
+            image: input image.
+
+        Returns:
+            detected keypoints, with maximum length of max_keypoints.
+        """
+
+
+class NewSuperpointDetector(NewDetector):
+    def __init__(self):
+        super(NewSuperpointDetector, self).__init__()
+        # init the opencv object
+        self._model = SuperPoint({"weights_path": MODEL_WEIGHTS_PATH}).cpu()
+        self._model.eval()
+
+    def apply(self, image: Image) -> Keypoints:
+        # Compute features.
+        image_tensor = torch.from_numpy(
+            np.expand_dims(image_utils.rgb_to_gray_cv(image).value_array.astype(np.float32) / 255.0, (0, 1))
+        )
+        with torch.no_grad():
+            model_results = self._model({"image": image_tensor})
+        torch.cuda.empty_cache()
+
+        # Unpack results.
+        coordinates = model_results["keypoints"][0].detach().cpu().numpy()
+        scores = model_results["scores"][0].detach().cpu().numpy()
+        keypoints = Keypoints(coordinates, scales=None, responses=scores)
+        # descriptors = model_results["descriptors"][0].detach().cpu().numpy().T
+
+        return keypoints
+
+
+@dask.delayed
+def detect(detector: NewDetector, image: Image) -> Keypoints:
+    return detector.apply(image)
```

Comment:
I plan to create functions like this for every module of GTSFM: detector, matcher, rotation averaging, etc. 

In the current design, we have `create_computation_graph` as a method of a detector class. In the `dask.delayed` call, we call another method of the class. This causes the whole class to be serialized and deserialized for every dask.delayed call. 
Curent code:
```
def create_computation_graph(self, image_graph: Delayed) -> Delayed:
        """Generates the computation graph for performing detection.
        Args:
            image_graph: computation graph for an image.
        Returns:
            Delayed task for detection on the input image.
        """
        return dask.delayed(self.detect)(image_graph)
```

In the current function, we delay the module, and pass it to the detect function. This gives us the potential to just serialize them once and scatter them to all the workers.

Reply:
Thanks Ayush. 
Do we need the function? Instead, can we make the model member a delayed object? So we could construct the class with a delayed (or future instance as you use it below) and reuse it in the method?

Reply:
I think the separation of dask related code into a function will be better. All the GTSFM processes will have no mention of dask at all. They just have to implement their API. I will make a new `dask_utils` which has one function for each process. What do you think?